src/dnnl/dnnl_graph_compiler.cpp

```diff
@@ -43,54 +41,120 @@
+    auto mod = JsonParser::parse(context, json, inputIds, outputIds, strides);
+    auto jmod = gc::JitModule::create(mod);
+
+    if (!static_cast<bool>(jmod)) {
+      auto err = jmod.takeError();
+      llvm::errs() << err;
+      llvm::consumeError(std::move(err));
+      std::string msg("Failed to create JitModule: ");
+      llvm::raw_string_ostream(msg) << err;
+      throw std::runtime_error(msg);
+    }
+
+    this->jmod = *jmod;
+  }
 
   void execute(dnnl_graph_compiler_tensor *inputs,
-               dnnl_graph_compiler_tensor *outputs) const;
+               dnnl_graph_compiler_tensor *outputs) const {
+    std::vector<MemRefWrapper> memRefs;
+    memRefs.reserve(inputIds.size() + outputIds.size());
+    for (auto &pair : {std::make_pair(&inputIds, inputs),
+                       std::make_pair(&outputIds, outputs)}) {
+      auto ids = pair.first;
+      auto tensors = pair.second;
+      for (size_t i = 0, n = ids->size(); i < n; i++) {
+        auto id = (*ids)[i];
+        dnnl_graph_compiler_tensor *tensor;
+
+        if (tensors[i].id == id) {
+          tensor = &tensors[i];
+        } else {
+          // The order of inputs/outputs may not match the function args order.
+          tensor = nullptr;
+          for (size_t j = 0; j < n; j++) {
+            if (tensors[j].id == id) {
+              tensor = &tensors[i];
+              break;
+            }
+          }
+          if (!tensor) {
+            throw std::invalid_argument("Tensor not found");
+          }
+        }
+
+        auto s = strides.find((*ids)[i]);
+        memRefs.emplace_back(tensor, s == strides.end() ? nullptr : &s->second);
+      }
+    }
+
+    llvm::SmallVector<void *> ptrs;
+    ptrs.reserve(memRefs.size());
+    for (auto &memRef : memRefs) {
+      ptrs.push_back(&memRef);
+    }
+    jmod->call(ptrs.data(), ptrs.size());
+  }
+
+private:
+  llvm::SmallVector<size_t> inputIds;
+  llvm::SmallVector<size_t> outputIds;
+  std::unordered_map<std::size_t, Strides> strides;
+  std::shared_ptr<gc::JitModule> jmod;
+
+  // C-compatible data wrapper -
+  // https://mlir.llvm.org/docs/TargetLLVMIR/#c-compatible-wrapper-emission
+  struct MemRefWrapper {
+    void *basePtr;
+    void *data;
+    int64_t offset = 0;
+    int64_t dimsAndStrides[2 * DNNL_MAX_NDIMS];
+
+    MemRefWrapper(dnnl_graph_compiler_tensor *tensor, const Strides *strides)
+        // We assume, that the data is aligned, thus basePtr == data.
+        : basePtr(tensor->data), data(tensor->data) {
+      if (tensor->ndims > DNNL_MAX_NDIMS) {
+        throw std::invalid_argument("Number of dimensions > DNNL_MAX_NDIMS");
+      }
+
+      std::copy(tensor->dims, tensor->dims + tensor->ndims, dimsAndStrides);
+      if (strides) {
+        std::copy(strides->begin(), strides->end(),
+                  dimsAndStrides + tensor->ndims);
+      } else {
+        std::fill(dimsAndStrides + tensor->ndims,
+                  dimsAndStrides + 2 * tensor->ndims, 0);
```

Comment:
Will it better to use the correct strides which could be inferred based on given `dims`?

Reply:
Sorry, didn't get the point. Could you please elaborate? The strides are read from the input json. If there are no strides, we just fill with zeroes here.

Reply:
I was referring to sth like this:
```cpp
int64_t stride = 1;
std::vector<int64_t> reversedStrides (tensor->ndims);
for (int64_t d = tensor->ndims - 1; d >0; d--) {
  reversedStrides.push_back(stride);
  stride *= tensor->dims[d];
}

std::reverse_copy(reversedStrides, reversedStrides+tensor->ndims, dimsAndStrides + tensor->ndims);
```

Reply:
Added.